crates/pop-common/src/sourcing/binary.rs

```diff
@@ -0,0 +1,253 @@
+	/// * `specified` - If available, a version explicitly specified.
+	/// * `available` - The available versions, used to check for those cached locally or the latest otherwise.
+	/// * `cache` - The location used for caching binaries.
+	pub fn resolve_version(
+		name: &str,
+		specified: Option<&str>,
+		available: &[impl AsRef<str>],
+		cache: &Path,
+	) -> Option<String> {
+		match specified {
+			Some(version) => Some(version.to_string()),
+			None => available
+				.iter()
+				.map(|v| v.as_ref())
+				// Default to latest version available locally
+				.filter_map(|version| {
+					let path = cache.join(format!("{name}-{version}"));
+					path.exists().then_some(Some(version.to_string()))
+				})
+				.nth(0)
+				.unwrap_or(
+					// Default to latest version
+					available.get(0).and_then(|version| Some(version.as_ref().to_string())),
+				),
+		}
+	}
+
+	/// Sources the binary.
+	///
+	/// # Arguments
+	/// * `release` - Whether any binaries needing to be built should be done so using the release profile.
+	/// * `status` - Used to observe status updates.
+	/// * `verbose` - Whether verbose output is required.
+	pub async fn source(
+		&self,
+		release: bool,
+		status: &impl Status,
+		verbose: bool,
+	) -> Result<(), Error> {
+		match self {
+			Self::Local { name, path, manifest, .. } => match manifest {
+				None => {
+					return Err(Error::MissingBinary(format!(
+						"The {path:?} binary cannot be sourced automatically."
+					)))
+				},
+				Some(manifest) => {
+					from_local_package(manifest, name, release, status, verbose).await
+				},
+			},
+			Self::Source { source, cache, .. } => {
+				source.source(cache, release, status, verbose).await
+			},
+		}
+	}
+
+	/// Whether any locally cached version can be replaced with a newer version.
+	pub fn stale(&self) -> bool {
+		// Only binaries sourced from GitHub release archives can currently be determined as stale
+		let Self::Source { source: GitHub(ReleaseArchive { tag, latest, .. }), .. } = self else {
+			return false;
+		};
+		latest.as_ref().map_or(false, |l| tag.as_ref() != Some(l))
+	}
+
+	/// Specifies that the latest available versions are to be used (where possible).
+	pub fn use_latest(&mut self) {
+		if let Self::Source { source: GitHub(ReleaseArchive { tag, latest, .. }), .. } = self {
+			if let Some(latest) = latest {
+				*tag = Some(latest.clone())
+			}
+		};
+	}
+
+	/// If applicable, the version of the binary.
+	pub fn version(&self) -> Option<&str> {
+		match self {
+			Self::Local { .. } => None,
+			Self::Source { source, .. } => match source {
+				Git { reference, .. } => reference.as_ref(),
+				GitHub(source) => match source {
+					ReleaseArchive { tag, .. } => tag.as_ref(),
+					SourceCodeArchive { reference, .. } => reference.as_ref(),
+				},
+				Archive { .. } | Source::Url { .. } => None,
+			},
+		}
+		.map(|r| r.as_str())
+	}
+}
+
+/// A descriptor of a remote repository.
+#[derive(Debug, PartialEq)]
+pub struct Repository {
```

Comment:
Debatable on whether this should have been moved as its currently only being used by the parachains crate. Probably fine. Questionable whether it should be in the binary module, apart from the package field which links to a buildable Rust package which _may_ be a binary.